### PR TITLE
making new cache for campaign objects

### LINKED ISSUE #2771: Campaign object caching
Implement a custom cache table to store the Campaign object returned by `dosomething_campaign_load`.
- [ ] Insert/update cache record upon node insert/update
- [ ] Return cache record within `dosomething_campaign_load`
- [ ] Add extra param to `dosomething_campaign_load` to bypass cache

### PR DESCRIPTION
**Pull request by katiecrane**
fixes #2771 

making new cache for campaign objects and storing them when `dosomething_campaign_load` is called

@weerd can you check on the `if ($public)` part of this chunk? does that need to be there if the campaign object is being pulled from the cache?

```
if ($cached_node = cache_get('ds_campaign_' . $node->nid, 'cache_dosomething_campaign') && $cache) { 
    // If this is accessed via API: 
    if ($public) { 
      // Add stats property. 
      dosomething_campaign_load_stats($campaign); 
      // Unset properties which shouldn't be public. 
      unset($campaign->variables); 
    } 
    return $campaign = $cached_node->data; 
  }
```

### CODE REVIEW

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc [hunk 1]
```diff
@@ -84,190 +86,205 @@
...
-    $campaign->scholarship = (int) $campaign->scholarship;
+  if ($cached_node = cache_get('ds_campaign_' . $node->nid, 'cache_dosomething_campaign') && $cache) {
+    // If this is accessed via API:
+    if ($public) {
+      // Add stats property.
+      dosomething_campaign_load_stats($campaign);
+      // Unset properties which shouldn't be public.
+      unset($campaign->variables);
+    }
+    return $campaign = $cached_node->data;
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc:103]: i think you would want to remove the variables from the `$cached_node->data`

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module [hunk 2]
```diff
@@ -1171,3 +1171,11 @@ function dosomething_campaign_get_campaign_runs($node) {
 
   return $campaign_runs;
 }
+
+/**
+ * Implements hook_flush_caches()
+ *
+ */
+function dosomething_campaign_flush_caches() {
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:1179]: you may want to add a flush_cache function for a specific campaign and update it on node updates

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc (hunk 1 above)
Comment (katiecrane) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc:103]: I'm not sure what you mean?

Comment (angaither) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc:103]: warning, fake code below

```
$cached_node = cache_get('whatever whatever)->data

if (public)
unset $cached_node->variables

return $cached_node
```

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module (hunk 2 above)
Comment (katiecrane) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:1179]: where do node updates happen?

**Comment by katiecrane** [comment]
@angaither I think this is done now? just pending approval from @weerd

**Comment by weerd** [comment]
Hey! Sorry @katiecrane was wrapped up in a few things end of last week; will check first thing on monday!

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module (hunk 2 above)
Comment (weerd) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:1179]: There should be some stuff in the `dosomething_campaign.module` file that has update related things.

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc [hunk 3]
```diff
@@ -84,190 +86,205 @@
...
+      // Same as the image_cover.
+      $campaign->image_header = $campaign->image_cover;
+    }
+
+    // Load Creator property.
+    dosomething_campaign_load_creator($campaign, $wrapper);
 
-  // Load Creator property.
-  dosomething_campaign_load_creator($campaign, $wrapper);
+    cache_set('ds_campaign_' . $campaign->nid, $campaign, 'cache_dosomething_campaign');
```
Comment (weerd) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc]: Might be good to not cache it forever, but add a time for how long to cache the node for.

Comment (angaither) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.helpers.inc]: @weerd there's a cache clear when the node gets updated, forever should be fine

**Comment by weerd** [comment]
@katiecrane also regarding the `$public` part, I don't think you'd need it if it's being pulled from cache.... Unless there's the possibility that when the first call for the campaign load happens and asks for public data, caches that and then on a load request of a cached version, passing `FALSE` for `$public`, could still load the cached version w/ public data exposed?

That all being said, this function is entirely _not_ used for the API except in the gallery (which I'm working ons witching out), so it's mostly just used internally ;)

**Comment by angaither** [comment]
> Unless there's the possibility that when the first call for the campaign load happens and asks for public data, caches that and then on a load request of a cached version, passing FALSE for $public, could still load the cached version w/ public data exposed?

@weerd yes, that's why I had her add that part. didn't want non-public data to be exposed in the cache. another way to go about this is to only cache it after public data has been removed.

**Comment by weerd** [comment]
@angaither Yeah, unless for some reason you need a version with the public data and the cached version has none?

But again, this is mostly used for internal stuff and no longer really being used for public API stuff other than gallery which is getting deprecated anyhoo...

**Comment by angaither** [comment]
@weerd got it. I don't think there's any harm in leaving it like it is though, right?

**Comment by weerd** [comment]
Yeah, I don't think so. Should be fine. I might add an expiration time to the `cache_set()` just in case for the time being, but other than that, should be fine (famous last words).

**Comment by katiecrane** [comment]
@angaither I added the timestamp back in and nothing broke. I think it's good now?

**Comment by angaither** [comment]
:+1:

**Pull request merged**